### ISSUE TITLE
Make non-literal indexing for ModuleList / ModuleDict work with ModuleInterface type in TorchScript

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#43737]
## 🚀 Feature

It is well-known that non-literal indexing for ModuleList / ModuleDict doesn't work (https://github.com/pytorch/pytorch/issues/16123) because from Zach (@zdevito): "TorchScript requires static typing and each module is a different type. So when trying to extract a module out of a list (or dict) with a variable, we do not know what type to assign it." (See internal discussion: https://fb.workplace.com/groups/329222650990087/permalink/715063122406036/?comment_id=715116715734010)

One idea to work around this is that if we know that all modules in the ModuleDict have the same known forward interface (let’s say all modules subclass from ModuleInterface), we might be able to make ModuleDict non-literal indexing work by giving all returned modules the ModuleInterface type.

## Motivation

In PyPer models, we use ModuleDict non-literal indexing extensively (e.g. https://www.internalfb.com/intern/diff/D23266836/?dest_fbid=1021488048305384&transaction_id=313476896604275). Not having ModuleDict non-literal indexing and having to enumerate all key-value pairs to get a specific module can cause perf regression, especially when we have a large number of modules in the ModuleDict (think hundreds or thousands).

## Pitch
API similar to the following:
```python
class ModuleA(ModuleInterface):
  ...

class ModuleB(ModuleInterface):
  ...

class TestModule(torch.nn.Module):
  def __init__(self):
    ...
    self.module_dict = torch.nn.ModuleDict({"mod_a": ModuleA(), "mod_b": ModuleB()})

  def forward(self, input):
    return self.module_dict["mod_a"](input) + self.module_dict["mod_b"](input)

scripted_module = torch.jit.script(TestModule())
scripted_module()
```

## Alternatives

The current workaround is to enumerate all key-value pairs of ModuleDict to get a specific module, but that's slow for large ModuleDict.

## Additional context

In PyPer models, we use ModuleDict non-literal indexing extensively (e.g. https://www.internalfb.com/intern/diff/D23266836/?dest_fbid=1021488048305384&transaction_id=313476896604275). The progress of converting PyPer models to TorchScript has been slowed down because of the lack of this feature in TorchScript.

cc @eellison 

cc @gmagogsfm

**Comment by ErenBalatkan** [comment]
This is one of the most common issues I face when trying to convert models to TorchScript. And i believe a quick temporary fix is possible, for example, allow us to work-around the issue by manually specifying type.

**Comment by SplitInfinity** [comment]
@ErenBalatkan Does #45716 help you? This allows you to index in to `ModuleDict` with a non-literal index if you provide a module interface as a type annotation. You can see examples in the unit tests in the PR.

**Comment by ErenBalatkan** [comment]
> @ErenBalatkan Does #45716 help you? This allows you to index in to `ModuleDict` with a non-literal index if you provide a module interface as a type annotation. You can see examples in the unit tests in the PR.

My problems are actually more related to ModuleLists right now, but if it also works with Lists, it could.

**Comment by gmagogsfm** [comment]
Closing. I believe this is fixed by #53410 and #45716.
